Implement exceptions for effect handlers using `Callback.register_exception`

kayceesrk opened the pull request:
This PR removes the pre-defined exceptions `Unhandled` and `Continuation_already_taken` used by effect handlers. Instead, we now declare these effects in the Effect module, and register them using `Callback.register_exception` to enable the runtime to raise these exceptions. 

This PR also replaces the exception `Unhandled` with `Unhandled_effect : 'a Effect.t -> exn`, which carries the unhandled effect. 

The PR supersedes #11419 and fixes #11185.

kayceesrk committed:
Use Callback.register_exception for exceptions in Effect module

kayceesrk committed:
Remove pre-defined exceptions for effect handlers

kayceesrk committed:
Bootstap.

This bootstrap removes the pre-defined exceptions Undefined and
Continuation_already_taken exceptions used by effect handlers.

dbuenzli commented:
Personally I find `Effect.Unhandled_effect` a bit long winded. I would go for `Effect.Unhandled`.

gasche commented on runtime/interp.c:
```diff
@@ -284,23 +284,25 @@ value caml_interprete(code_t prog, asize_t prog_size)
+      (code_t)raise_unhandled_effect_code;
+    Closinfo_val(raise_unhandled_effect_closure) = Make_closinfo(0, 2);
+    raise_unhandled_effect = raise_unhandled_effect_closure;
+    caml_register_generational_global_root(&raise_unhandled_effect);
```
Can you explain to a newcomer why this black magic around `raise_unhandled_effect` is necessary? (The explanation could be added as a comment in the code.)

I'm not sure what parts of the bytecode are allowed to directly raise exceptions (apparently `caml_raise_continuation_already_taken()` below is fine, and there are existing examples with `caml_raise_stack_overflow` for example), and which are not / why. Here what's going on seems very complex compared to the semantics we want, or even to the handling of unhandled exceptions in

https://github.com/ocaml/ocaml/blob/37df9347658338f488e74780d9e1b0cf9eecc4fc/runtime/interp.c#L972-L977

kayceesrk replied:
Essentially, `raise_unhandled_effect` is necessary when we do not find a matching effect handler in our stack of handlers. The search had proceeded by checking one effect handler on the stack at a time. When we reach the bottom of the stack of effect handlers and find that there is no parent stack, we have to get the control back to the point of `perform` (`do_resume`) and then raise the exception. `raise_unhandled_effect` is the closure that raises the exception immediately after the control returns to the point of perform. 

I'll write a clarifying comment in the code.

dra27 commented:
I'm not sure that an example can actually be concocted which would trigger this, but it reads slightly strangely that the `Effect` module initialises a runtime structure (in this case the two exceptions in the named values table). In other words, is it possible to have a program which could lead to `caml_raise_unhandled_effect` where the exception has not been registered, thinking of full stdlib replacements, etc., or linking without `-linkall`. Is the registration something which ought to be in `CamlinternalEffect` and explicitly called as part of the startup code on the OCaml side?

favonia commented:
Typo in the Git commit message of https://github.com/ocaml/ocaml/pull/11423/commits/37df9347658338f488e74780d9e1b0cf9eecc4fc: "`Undefined`"

xavierleroy commented:
I won't review the implementation until we all agree on the interface and the documentation.  +1 for @dbuenzli's suggestions, especially the suggestion to shorten `Effect.Unhandled_effect` to `Effect.Unhandled`.

xavierleroy commented on testsuite/tests/effects/reperform.reference:
```diff
@@ -19,4 +19,4 @@
  !4]
  !5]
-Unhandled
+Stdlib.Effect.Unhandled_effect(_)
```
This `(_)` is unfortunate and reduces the usefulness of including the effect in the exception.  It would be 100% better to also print the name of the effect !  For this we'll need some ad-hoc code in the exception printer(s), I'm afraid.

kayceesrk replied:
Good point. I see that there is similar code in `printexc.ml`.

favonia commented:
As the person who originally requested the renaming of `Unhandled` (unscoped) to `Unhandled_effect`, I am satisfied with `Effect.Unhandled`. :smiley:

kayceesrk commented:
> Is the registration something which ought to be in CamlinternalEffect and explicitly called as part of the startup code on the OCaml side?

It would be odd to not find the exception when using `caml_named_value`. Currently, there is an assertion to check that the exception exists.

https://github.com/ocaml/ocaml/blob/37df9347658338f488e74780d9e1b0cf9eecc4fc/runtime/fiber.c#L669-L674

If we think that there is a possibility of these exceptions not being registered (with stdlib replacements), then it may be easiest to replace this assertion with a check that `exn` is not null and fail with a `caml_fatal_error` if it is null.

Octachron commented:
@dra27: If I am not mistaken, the exceptions can only be raised by OCaml code that links Effect, thus it seems fine to do the initialization only when linking the `Effect` module. If a stdlib replacement exposes the runtime effect primitive functions by itself without linking the Effect module, it should also take care to properly register the exceptions itself.

lthls commented:
> @dra27: If I am not mistaken, the exceptions can only be raised by OCaml code that links Effect, thus it seems fine to do the initialization only when linking the `Effect` module.

That's unfortunately not the case. Defining a new effect only needs to reference the interface, not the implementation, and the same applies for performing an effect (`Effect.perform` is an external declaration). So you can have code performing effects that has does not link with the `Effect` module.
Writing effect handlers currently needs to go through the implementation of `Effect`, but even that might change if we introduce support in the language itself.

Overall, I consider that if there is any amount of initialisation code in a module, it should be compiled with `-linkall`. This is annoying for Stdlib modules as that means it's almost impossible to not link them even if they're not used, so I'd advise one of the two following alternatives:
* Switch back to builtin exceptions (you can still rebind a builtin exception in `Effect`, using `type exn += Unhandled = Unhandled_effect` for instance)
* Make the `Effect` module an independent library

kayceesrk commented:
> Defining a new effect only needs to reference the interface, not the implementation, and the same applies for performing an effect (Effect.perform is an external declaration). So you can have code performing effects that has does not link with the Effect module.

This seems like an adversarial case. Is it possible to write programs which will perform effects without linking to the Effect module? In other words, is this problem encountered by programmers who are not actively trying to use effects without linking to the effect module. We should certainly address this issue if the programmer is not trying to do something clever but encounters this issue. 

> so I'd advise one of the two following alternatives:
> 
> * Switch back to builtin exceptions (you can still rebind a builtin exception in Effect, using type exn += Unhandled = Unhandled_effect for instance)
> * Make the Effect module an independent library
> 

What do you think about the following options? 

1. We raise `Failure "Effect module not linked"` if the runtime is not able to find the exceptions. 
2. Separating the initialisation code to `camlinternalEffect`?

xavierleroy commented:
Again we're discussing implementation before we've agreed on the API...

> We raise Invalid_argument "Effect module not linked" if the runtime is not able to find the exceptions.

That's what otherlibs/unix does for the Unix_error exception.  A fatal error is also possible, see
https://github.com/ocaml/ocaml/blob/ef376dd4160f53970a21b6129cb83d460b97614b/runtime/fail_nat.c#L204-L218

Please forget about the implementation and discuss the API (incl. documentation comments).

kayceesrk commented:
> Again we're discussing implementation before we've agreed on the API...

Do we agree that changing the exception `Unhandled` to `Effect.Unhandled : 'a Effect.t -> exn` is a good idea? Are there other API-related questions that we are missing? 

There are a few (potentially non-controversial) tasks left for improving the PR:

* Rename `Effect.Unhandled_effect` to `Effect.Unhandled`. 
* Implement ad-hoc printer for effect values.
* Improve API docs based on these comments: https://github.com/ocaml/ocaml/pull/11419#pullrequestreview-1034329747

Octachron commented:
I think that renaming the exception to `Effect.Unhandled` is definitively a good idea.

Thinking a bit about the `Effect` module, I was wondering about the long term expectation for this module. As far as I understand, we expect to add support in the language itself for most of the primitives exposed by the module. What  will happen to this module at this point? Will it contain only some utility functions? Or might it be completely removed at some point?

If the later option is a possibility, it might make sense to expose the effect module as a library. This has the advantage of making it clearer that the module is an experimental API that does not come with the backward-compatibility guarantee of the standard library. Moreover,  integrating a library in the standard library is both simpler and faster than removing a module from the standard library.

dbuenzli commented:
> This has the advantage of making it clearer that the module is an experimental API that does not come with the backward-compatibility guarantee of the standard library. Moreover, integrating a library in the standard library is both simpler and faster than removing a module from the standard library.

It also entails quite a bit of bureaucracy. In general the whole process around the stdlib remains a bit unrealistic in my opinion, especially for something like the 5.0 transition.

You can think as much as you want about design, design errors still occur and most of the time you only get to know about them once it's in the wild and the stuff gets used in anger.

The current process entails that if a design error gets made it is cast in stone for the next 20 years which is not so great. I think the stdlib should have a notion of `@experimental` when new APIs or even functions are introduced which would allow for corrections for as long as the tag applies. As an end user, you have been warned.

Personally I would even flag all of the entry points introduced specifically for `5.0` as such (e.g. for the little I have looked into I'm already [not really happy about](https://github.com/ocaml/ocaml/issues/11074) the `Domain` interface).

xavierleroy commented on stdlib/effect.mli:
```diff
@@ -19,10 +19,18 @@
 type _ t = ..
 (** The type of effects. *)
 
+exception Unhandled_effect : 'a t -> exn
+(** Exception raised when an effect is performed and there is no matching
+    handler for that effect. The argument is the unhandled effect. *)
+
+exception Continuation_already_taken
+(** Exception raised when a continuation is resumed with a [continue] or
+    [discontinue] but the continuation has already been resumed. *)
+
```
Did we agree that the exception should be `Effect.Unhandled` instead of `Effect.Unhandled_effect`, or are we still on the fence?

Also, @dbuenzli's comments on the doc strings in #11419 are still relevant and should be addressed.

xavierleroy commented on manual/src/refman/extensions/effects.etex:
```diff
@@ -601,13 +601,13 @@ fast.
+handler, then an "Unhandled_effect" exception is raised at the point of the
 corresponding "perform". For example, in the previous example, "bar" does not
-handle the effect "F". Hence, we will get an "Unhandled" exception when we run
-"bar".
+handle the effect "F". Hence, we will get an "Unhandled_effect F" exception
+when we run "bar".
 
 \begin{caml_example}{verbatim}
-try bar () with Unhandled -> "Saw unhandled exception"
+try bar () with Unhandled_effect F -> "Saw Unhandled_effect exception"
```
This may need to be updated again (`Unhandled_effect` -> `Effect.Unhandled`).

kayceesrk replied:
Fixed by https://github.com/ocaml/ocaml/pull/11423/commits/e402a83501ddd36994d268ffe85c146d518857a0.

xavierleroy reviewed (commented):
Let's try to agree on the API, shall we?

jonludlam committed:
Rename Effect.Unhandled_effect -> Effect.Unhandled

jonludlam committed:
Handle effect names in Effect.Unhandled exception stringify

It seemed sensible not to duplicate the logic to turn sum types into
strings that is currently in use for exceptions. However, since it's
not exposed in Printexc, and the exception being matched is in
effect.ml, which depends on Printexc, I've opted for matching the
string "Stdlib.Effect.Unhandled" and assuming the single argument is
the effect.

jonludlam committed:
Documentation updates for Effects

kayceesrk committed:
Avoid string comparison for identifying Unhandled exception in Printexc

kayceesrk commented:
Thanks @jonludlam for the help. With the latest commits, the outstanding tasks mentioned in https://github.com/ocaml/ocaml/pull/11423#issuecomment-1182797192 are complete.

favonia commented:
It seems there are a few additional details of the API that were implicitly agreed upon: (Or, at very least, I did not see any objections.)
1. It should be possible to use effects without the standard library. (Suggested by @dra27 but not documented.)
2. There should _not_ be any top-level aliases of the exception `Effect.Unhandled`.

The reason I brought up (2) is that it seems to be the only reason to favor exception registration over built-in exceptions. As for (1), if the current PR is tracking this API design, perhaps `Continuation_already_taken` and the registration of `Unhandled` and `Continuation_already_taken` should all be moved to the module `CamlinternalEffect`?

Personally, I also like @dbuenzli's suggestion to mark new things experimental. Is this part of the API discussion? If so, I support marking both `Effect` and `Domain` (especially `Domain`) experimental in the sense that backward compatibility is not guaranteed.

xavierleroy commented:
> It should be possible to use effects without the standard library. (Suggested by @dra27 but not documented.)

Just to clarify: I do not consider this necessary at all.  First, anyone can build on top of the standard library, there is seldom a need to replace it entirely.  Second, a stdlib replacement can register exceptions itself, like the stdlib would do.

Built-in exceptions and `Camlinternal*` modules are best avoided if we can: built-in exceptions because they pollute the scope and complicate the bootstrap, and `Camlinternal*` modules because they complicate the structure of the stdlib.  Here, we can avoid both.  Let's do that.

xavierleroy reviewed (commented):
I read `stdlib/effect.mli` again, and the API and doc strings look just fine to me.  Thanks!  @favonia @dbuenzli @fpottier and others: if you have reservations about the current API and documentation, now is the time to tell.

I started looking at the implementation and I have a number of suggestions to make, probably tomorrow.

dbuenzli commented on runtime/caml/fiber.h:
```diff
@@ -168,14 +168,15 @@ struct c_stack_link {
- *   switch back to the continuation stack to raise the Unhandled exception in
- *   in the case no parent is left. Otherwise the instruction switches to the
- *   parent stack and executes the handle_effect function for that parent stack.
+ *   switch back to the continuation stack to raise the Unhandledi_effect
```
Typo `Unhandledi_effect` -> `Effect.Unhandled`

favonia replied:
@Octachron It seems you missed this one. :slightly_smiling_face:

Octachron replied:
Indeed, the typo escaped the renaming, this is fixed now.

dbuenzli reviewed (commented):
Docs and names look good. Thank @kayceesrk.

favonia commented on stdlib/effect.mli:
```diff
@@ -19,10 +19,18 @@
 (** [perform e] performs an effect [e].
 
-    @raise Unhandled if there is no active handler. *)
+    @raise Unhandled e if there is no handler for [e]. *)
```
```suggestion
    @raise Unhandled if there is no handler for [e]. *)
```
Unfortunately `@raise Unhandled e if...` will be parsed as "`@raised`", "`Unhandled`", and "`e if ...`". On the other hand, the `[e]` was already mentioned in the beginning of the documentation, so I suppose it's okay to omit the first `e` in this sentence?

favonia replied:
@Octachron I think you might have missed this.

Octachron replied:
I agree that omitting the exception payload is more in line with the rest of the documentation. This is fixed now.

favonia commented on stdlib/printexc.ml:
```diff
@@ -52,6 +52,19 @@ let use_printers x =
     | [] -> None in
   conv (Atomic.get printers)
 
+let walk_sum_type x =
```
_(This is about the implementation.)_ I personally prefer the name `extensible_variant` over `sum_type` because the printer here does not work for all variant types (and in OCaml documentation the term "sum types" is not used). This comment also applies to the function `string_of_sum_type` a few lines below.

favonia commented on stdlib/printexc.ml:
```diff
@@ -61,14 +74,13 @@ let to_string_default = function
   | Undefined_recursive_module(file, line, char) ->
       sprintf locfmt file line char (char+6) "Undefined recursive module"
+  | CamlinternalEffect.Unhandled e ->
+      sprintf "Stdlib.Effect.Unhandled(%s)"
```
```suggestion
      sprintf "Unhandled effect: %s"
```
Given that we are providing more user-friendly messages for other built-in exceptions, I think this would look nicer.

Octachron replied:
Let's keep the usual exception message style for now. We can decide to improve the message at any later point (contrarily to the exception names).

favonia commented on stdlib/printexc.ml:
```diff
@@ -61,14 +74,13 @@ let to_string_default = function
+  | CamlinternalEffect.Unhandled e ->
+      sprintf "Stdlib.Effect.Unhandled(%s)"
+        (string_of_sum_type (walk_sum_type (Obj.repr e)))
   | x ->
```
Maybe there should also be a special case for `Effect.Continuation_already_taken`. More precisely,
```ocaml
  | Effect.Continuation_already_taken -> "Continuation already taken"
```
Actually, should it be `Continuation_already_resumed` (and thus "Continuation already resumed") for perfect consistency with the API documentation? Also https://github.com/ocaml/ocaml/issues/11422 might be related.

kayceesrk replied:
Indeed. `Continuation_already_resumed` feels more appropriate than "taken".

favonia reviewed (commented):
I noticed that in `effects.etex`, `Unhandled` is qualified (as `Effect.Unhandled`) but `Continuation_already_taken` is not, even though both are defined at the same place now. Otherwise, everything else looks great. @kayceesrk Thank you. :slightly_smiling_face:

kayceesrk commented:
@xavierleroy 

> Camlinternal* modules because they complicate the structure of the stdlib

I needed to introduce the `CamlInternalEffect` module to break the cyclic dependency between the `Printexc` and `Effect` modules. Agree that it would be good to avoid the introduction of a new `CamlInternal*` module if there is way around this. 

(Apologies for the slow response on this thread. It is the first week of teaching in the new semester for me.)

xavierleroy commented:
> I needed to introduce the CamlInternalEffect module to break the cyclic dependency between the Printexc and Effect modules.

Well, an alternative is to not change `Printexc` at all, just register an appropriate printer from within `effect.ml`, something like:
```
let print_exceptions = function
  | Unhandled e -> Some ("Effect.Unhandled " ^ Obj.Extension_constructor.(name (of_val e)))
  | Continuation_already_taken -> Some "Effect.Continuation_already_taken"
  | _ -> None

let _ =
  Printexc.register_printer print_exceptions
```

kayceesrk commented:
There are a number of issues with the current implementation.

## Exception names not registered

The following program crashes since exception names may not registered. This was pointed out by @lthls in https://github.com/ocaml/ocaml/pull/11423#issuecomment-1181695853. I didn't understand the issue originally. Now I see it. 

```ocaml
$ cat test.ml
open Effect
type _ t += E : unit t
let _ = perform E
$ ../_install/bin/ocamlopt.opt test.ml
$ ./a.out 
Segmentation fault (core dumped) 
```

whereas this program doesn't:

```ocaml
$ cat test2.ml
open Effect
type _ t += E : unit t
let _ = Effect.Deep.try_with (* links the Effect module *)
let _ = perform E
```

In the above program, `Effect` module gets linked and the initialisation is performed, which registers the exceptions. This is a **must-fix** issue. 

## More work is necessary for ad-hoc printing

The default uncaught exception handler does not print the effect name in `Effect.Unhandled` if the `Printexc` module is not linked. Consider the program `test2.ml` above, which does not link the `Printexc` module:

```ocaml
$ ocamlopt test2.ml
$ ./a.out
Fatal error: exception CamlinternalEffect.Unhandled(_)
```

If the Printexc module is linked, then we see the effect name printed. Consider the example below:

```ocaml
$ cat test3.ml
open Effect
type _ t += E : unit t
let _ = Effect.Deep.try_with (* links the Effect module *)
let _ = Printexc.to_string (* links the Printexc module *)
let _ = perform E
$ ocamlopt test3.ml
$ ./a.out 
Fatal error: exception Stdlib.Effect.Unhandled(Test3.E)
```

It is a bit smelly that linking the `Printexc` module is necessary for printing the effects.

xavierleroy commented:
There's an easy solution: in effect.mli, declare `perform` as a `val`, not an `external`.  This way, any use of `Effect.perform` will cause the Effect module to be linked and its initialization code to be performed.  If said initialization code registers an exception printer, the Printexc module will also be linked.  Easy peasy.

xavierleroy committed:
Guard against missing registration for exceptions declared in Effect

In this case, produce a somewhat clean fatal error if these exceptions
are raised.

xavierleroy commented:
I played with the various alternatives mentioned above.
- Declaring `perform` as `val` in effect.mli works but makes stack backtraces worse.
- Registering a printer in `Effect` instead of having dedicated code in `Printexc` to handle `Effect.Unhandled` works too, but the unhandled effect is printed with fewer details (just the name of the effect, not the arguments).

Concerning the crash when raising exceptions that have not been registered, here is a simple fix: https://github.com/xavierleroy/ocaml/commit/424c92d1ee9103fa936aac483fb979d4f1017d6a .  Just like for the "out of bounds" exception, if the exception has not been registered, we produce a fatal error that looks like an uncaught exception error.

Octachron committed:
Move Unhandled_effect printers to Effect

Octachron commented:
Concerning the less detailed printer, this can be solved by using the same  `Obj` printer than `Printexc`, isn't it?

I have the impression that if we register the `Unhandled` exception printer augmented with this obj printer in the Effect module and raise a fatal error if the `Effect` module was not linked, we end up with a quite good compromise:

- users that never linked the Effect module (because they never installed any effect handler) will see the fatal error
- linking the `Effect` module installs the more informative `Unhandled_effect` exception printer.
- types and effect-related exceptions are confined to the Effect module
-  there are no new `camlinternal*` module.

Octachron commented:
Here is an implementation of my proposal to move the exception printer inside the `Effect` module: https://github.com/Octachron/ocaml/commit/78f2d2ae7007d1ed11ffdba1970d34c720652edb .

Octachron commented:
@kayceesrk , gentle ping: if possible I would prefer to have this change in the first beta.

Do you have any opinions on the commit that @xavierleroy and me are proposing?
(If time is sparse on your side, I could take care of the rebasing?)

favonia commented:
Assuming rebasing will happen, let me summarize a few comments I made that could affect the final API to speed things up:
1. `Continuation_already_resumed` looks better than `Continuation_already_taken` (and @kayceesrk seems to agree).
2. Printer should take care of `Continuation_already_resumed/taken` too (as of other built-in exceptions).
3. For built-in exceptions, the output messages have been short English phrases (`"Undefined recursive module"`) instead of raw constructor names (`"Undefined_recursive_module"`). So I prefer `"Unhandled effect: %s"` over `"Stdlib.Effect.Unhandled(%s)"`.

kayceesrk commented:
@Octachron the commit looks good to me. It would be great if you can rebase this change and push the PR towards a merge. I'm unfortunately quite busy before ICFP. Thank you.

Octachron committed:
Rename Continuation_already_taken to Continuation_already_resumed

The documentation was already using "resumed" rather than taken which is
a good sign that "Continuation_already_resumed" is clearer.

Octachron committed:
test: perform an effect without linking Effect

Octachron commented:
I have updated and rebased the PR and added @xavierleroy's commit and mine to the branch.
I have also renamed the `Continuation_already_taken` exception to `Continuation_already_resumed` to align the exception name with the documentation.

@favonia: After this PR neither the `Continuation_already_taken` exception nor `Unhandled` exception are built-in exceptions. It thus seems reasonable to me to keep them printed as regular exceptions.

Octachron committed:
typo

favonia commented:
> After this PR neither the `Continuation_already_taken` exception nor `Unhandled` exception are built-in exceptions. It thus seems reasonable to me to keep them printed as regular exceptions.

I think the difference between "a built-in exception" and "an exception defined in the standard library" is very small. It's only meaningful to people who build compilers or alternative standard libraries. Therefore, I still feel it is better to include all (public) exceptions defined in the standard library. That said, I do not have a strong opinion because the output will be clear either way. I see that there are some unaddressed comments about the documentation. Other than those, I am very satisfied with the current PR and hope it can be shipped with the next alpha or the beta! :smiley:

Octachron committed:
Fix flambda test

Octachron committed:
Tweak the Effect module documentation

Use "@raise Unhandled" rather than "@raise Unhanded [e]"

kayceesrk commented:
The PR as it stands looks good to me. The only point of contention is the choice between the strings `"Unhandled effect: %s"` and `"Stdlib.Effect.Unhandled(%s)"`. This can be changed in a future version without breaking code. 

I realised I can't approve the PR since I'd opened it! @Octachron feel free to merge when you are happy.

xavierleroy reviewed (approved):
Approving on behalf of @kayceesrk and @Octachron , who both wrote the code and reviewed the other's code.

xavierleroy commented:
There's a bootstrap in the middle, so I'd rather let @Octachron handle the final merge.

Octachron commented:
Merged and cherry-picked by hand to 5.0 .

Octachron commented:
The `reperform` test appears to be failing on `Windows-mingw64`, I will investigate the issue.
